pallets/nfts/src/tests.rs

```diff
@@ -2135,59 +2981,260 @@ fn cancel_approval_works_with_force() {
 		);
 	});
 }
 
+#[test]
+fn clear_collection_approvals_works() {
+	new_test_ext().execute_with(|| {
+		let balance = 100;
+		let collection_id = 0;
+		let delegate_1 = account(3);
+		let delegate_2 = account(4);
+		let item_id = 42;
+		let owner = account(1);
+
+		// Origin checks for the `clear_collection_approvals`.
+		for origin in [root(), none()] {
+			assert_noop!(
+				Nfts::clear_collection_approvals(origin, collection_id, 0),
+				BadOrigin.with_weight(WeightOf::<Test>::clear_collection_approvals(0))
+			);
+		}
+		// Origin checks for the `force_clear_collection_approvals`.
+		for origin in [RuntimeOrigin::signed(owner.clone()), none()] {
+			assert_noop!(
+				Nfts::force_clear_collection_approvals(origin, owner.clone(), collection_id, 0),
+				BadOrigin.with_weight(WeightOf::<Test>::clear_collection_approvals(0))
+			);
+		}
+
+		assert_ok!(Nfts::force_create(
+			RuntimeOrigin::root(),
+			owner.clone(),
+			default_collection_config()
+		));
+		assert_ok!(Nfts::force_mint(
+			RuntimeOrigin::signed(account(1)),
+			collection_id,
+			item_id,
+			owner.clone(),
+			default_item_config()
+		));
+
+		for (origin, maybe_owner) in [
+			// Parameters for `clear_collection_approvals`.
+			(root(), Some(owner.clone())),
+			// Parameters for `force_clear_collection_approvals`.
+			(RuntimeOrigin::signed(owner.clone()), None),
+		] {
+			Balances::make_free_balance_be(&owner, balance);
+
+			assert_ok!(Nfts::approve_collection_transfer(
+				RuntimeOrigin::signed(owner.clone()),
+				collection_id,
+				delegate_1.clone(),
+				None
+			));
+			assert_ok!(Nfts::approve_collection_transfer(
+				RuntimeOrigin::signed(owner.clone()),
+				collection_id,
+				delegate_2.clone(),
+				None
+			));
+			// Removes zero collection approval.
+			assert_eq!(
+				clear_collection_approvals(origin.clone(), maybe_owner.clone(), collection_id, 0),
+				Ok(Some(WeightOf::<Test>::clear_collection_approvals(0)).into())
+			);
+			assert_eq!(Balances::free_balance(&owner), balance - 2);
+			assert_eq!(
+				CollectionApprovals::<Test>::iter_prefix((collection_id, owner.clone())).count(),
+				2
+			);
+			assert!(!events().contains(&Event::<Test>::ApprovalsCancelled {
+				collection: collection_id,
+				item: None,
+				owner: owner.clone(),
+			}));
+
+			// Partially removes collection approvals.
+			assert_eq!(
+				clear_collection_approvals(origin.clone(), maybe_owner.clone(), collection_id, 1),
+				Ok(Some(WeightOf::<Test>::clear_collection_approvals(1)).into())
+			);
+			assert_eq!(Balances::free_balance(&owner), balance - 1);
+			assert_eq!(
+				CollectionApprovals::<Test>::iter_prefix((collection_id, owner.clone())).count(),
+				1
+			);
+
+			// Successfully removes all collection approvals. Only charges post-dispatch weight for
+			// the removed approvals.
+			assert_eq!(
+				clear_collection_approvals(origin.clone(), maybe_owner.clone(), collection_id, 10),
+				Ok(Some(WeightOf::<Test>::clear_collection_approvals(1)).into())
```

Comment:
Note: Post-dispatch weight is charged correctly with only the number of collection approvals. In this case, `limit = 10` is provided but we only charge for `1` collection approval removed.